src/ucp/wireup/wireup_cm.c

```diff
@@ -0,0 +1,272 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2019.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include "wireup_cm.h"
+#include <ucp/core/ucp_ep.inl>
+#include <ucp/wireup/wireup.h>
+#include <ucp/wireup/wireup_ep.h>
+
+
+static unsigned
+ucp_cm_ep_init_flags(const ucp_worker_h worker, const ucp_ep_params_t *params)
+{
+    if (!ucp_worker_sockaddr_is_cm_proto(worker)) {
+        return 0;
+    }
+
+    if (params->field_mask & UCP_EP_PARAM_FIELD_SOCK_ADDR) {
+        return UCP_EP_INIT_CM_WIREUP_CLIENT;
+    }
+
+    if (params->field_mask & UCP_EP_PARAM_FIELD_CONN_REQUEST) {
+        return UCP_EP_INIT_CM_WIREUP_SERVER;
+    }
+
+    return 0;
+}
+
+static ucs_status_t ucp_cm_preinit_client_lanes(ucp_ep_h ucp_ep)
+{
+    ucp_worker_h worker        = ucp_ep->worker;
+    ucp_ep_config_key_t key    = ucp_ep_config(ucp_ep)->key;
+    uint64_t addr_pack_flags   = UCP_ADDRESS_PACK_FLAG_DEVICE_ADDR |
+                                 UCP_ADDRESS_PACK_FLAG_IFACE_ADDR  |
+                                 /* Disable HW AMO because ucp_ep lanes are
+                                  * bound to specific device which may be
+                                  * different form worker->atomic_tls */
+                                 UCP_ADDRESS_PACK_FLAG_DISABLE_HW_AMO;
+    ucp_wireup_ep_t *wireup_ep = ucp_ep_get_cm_wireup_ep(ucp_ep);
+    void *ucp_addr;
+    size_t ucp_addr_size;
+    ucp_unpacked_address_t unpacked_addr;
+    uint8_t *addr_indices;
+    ucp_lane_index_t lane_idx;
+    ucs_status_t status;
+
+    ucs_assert_always(wireup_ep != NULL);
+
+    /* Construct local dummy address for lanes selection taking an assumption
+     * that server has the transports which are the best from client's
+     * perspective. */
+    status = ucp_address_pack(worker, NULL,
+                              ucp_context_tl_bitmap(worker->context,
+                                                    wireup_ep->dev_name),
+                              addr_pack_flags, NULL, &ucp_addr_size, &ucp_addr);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    status = ucp_address_unpack(worker, ucp_addr, addr_pack_flags,
+                                &unpacked_addr);
+    if (status != UCS_OK) {
+        goto free_addr;
```

Comment:
you need to add other `goto` label, because here you need to free `ucp_addr`.
`goto free_addr` would try to free `unpacked_addr.address_list`, which is not safe.
In general freeing `NULL` is safe, but we are not sure that `unpacked_addr.address_list` is `NULL` if `ucp_address_unpack` fails

Reply:
refactored a bit